nautobot_data_validation_engine/custom_validators.py

```diff
@@ -118,6 +128,140 @@ def clean(self):
+
+
+def get_data_compliance_rules_map():
+    """Generate a dictionary of audit rulesets associated to their models."""
+    compliance_rulesets = {}
+    for validators in registry["plugin_custom_validators"].values():
+        for validator in validators:
+            if is_data_compliance_rule(validator):
+                compliance_rulesets.setdefault(validator.model, [])
+                compliance_rulesets[validator.model].append(validator)
+
+    return compliance_rulesets
+
+
+def get_classes_from_git_repo(repo: GitRepository):
+    """Get list of DataComplianceRule classes found within the custom_validators folder of the given repo."""
+    ensure_git_repository(repo, head=repo.current_head)
+    class_list = []
+    for importer, discovered_module_name, _ in pkgutil.iter_modules([f"{repo.filesystem_path}/custom_validators"]):
+        if discovered_module_name in sys.modules:
+            del sys.modules[discovered_module_name]
+        module = importer.find_module(discovered_module_name).load_module(discovered_module_name)
+        for _, complance_class in inspect.getmembers(module, is_data_compliance_rule):
+            class_list.append(complance_class)
+    return class_list
+
+
+class ComplianceError(ValidationError):
+    """A compliance error is raised only when an object fails an compliance check."""
+
+
+class DataComplianceRule(CustomValidator):
+    """Class to handle a set of validation functions."""
+
+    class_name: Optional[str] = None
+    model: str
+    result_date: timezone
+    enforce = False
+
+    def __init__(self, obj):
+        """Initialize an DataComplianceRule object."""
+        super().__init__(obj)
+        self.class_name = self.class_name or self.__class__.__name__
+        self.result_date = timezone.now()
+
+    def audit(self):
+        """Not implemented.  Should raise an ComplianceError if an attribute is found to be invalid."""
+        raise NotImplementedError
+
+    def mark_existing_attributes_as_valid(self, exclude_attributes=None):
+        """Mark all existing fields (any that were previously created) as valid=True."""
+        instance = self.context["object"]
+        if not exclude_attributes:
+            exclude_attributes = []
+        attributes = (
+            DataCompliance.objects.filter(
+                compliance_class_name=self.class_name,
+                content_type=ContentType.objects.get_for_model(instance),
+                object_id=instance.id,
+            )
+            .exclude(validated_attribute__in=["all"] + exclude_attributes)
+            .values_list("validated_attribute", flat=True)
+        )
+        for attribute in attributes:
+            self.compliance_result(message=f"{attribute} is valid.", attribute=attribute)
+
+    def clean(self):
+        """Override the clean method to run the audit function."""
+        try:
+            self.audit()
+            self.mark_existing_attributes_as_valid()
```

Comment:
So an example of how it's currently working: let's say I have a `DataComplianceRule` that checks attributes `foo` and `bar`.  If during the first `audit` run there are no issues with either attribute, the only `DataCompliance` object created would be with attribute `all` saying that the given object is valid.  

Then I realize the logic for checking `foo` was wrong, so I update that and rerun the `DataComplianceRule`, which then marks `foo` as invalid and creates a `DataCompliance` object for `foo` being invalid and updates the `all` attribute result to say the object is not valid.  

Then if I went and fixed the issue in the value for `foo` for the given object, the `DataCompliance` objects would be updated for `foo` to be valid and for `all` to be valid.  In this scenario, `bar` would never have a `DataCompliance` object created for this object since it's never been invalid.  Only attributes that were at one point invalid per the `DataComplianceRule` will have `DataCompliance` objects created for them.